SmartBadge/lib/ext/bt/ble_central.py

```diff
@@ -0,0 +1,198 @@
+
+        elif event == _IRQ_GATTC_READ_RESULT:
+            # A read completed successfully.
+            conn_handle, value_handle, char_data = data
+            if conn_handle == self._conn_handle and value_handle == self._value_handle:
+                self._update_value(char_data)
+                if self._read_callback:
+                    self._read_callback(self._value)
+                    self._read_callback = None
+
+        elif event == _IRQ_GATTC_NOTIFY:
+            conn_handle, value_handle, notify_data = data
+            if conn_handle == self._conn_handle and value_handle == self._value_handle:
+                self._update_value(notify_data)
+                if self._notify_callback:
+                    self._notify_callback(self._value)
+
+
+    # Returns true if we've successfully connected and discovered characteristics.
+    def is_connected(self):
+        return self._conn_handle is not None and self._value_handle is not None\
+        and self._value_handle_rx is not None and self._value_handle_tx is not None
+
+    # Find a device advertising the environmental sensor service.
+    def scan(self, callback=None):
+        # If starting a new scan delete the prev. found devices
+        if self._scan_complete:
+            self._detected_devices = []
+            self._scan_complete = False
+
+        self._addr_type = None
+        self._addr = None
+        self._scan_callback = callback
+        self._ble.gap_scan(2000, 30000, 30000)
+
+    def detected_devices(self):
+        return self._detected_devices
+
+    # Connect to the specified device (otherwise use cached address from a scan).
+    def connect(self, addr_type=None, addr=None, callback=None):
+        self._addr_type = addr_type or self._addr_type
+        self._addr = addr or self._addr
+        self._conn_callback = callback
+        if self._addr_type is None or self._addr is None:
+            return False
+        self._ble.gap_connect(self._addr_type, self._addr)
+        return True
+
+    # Disconnect from current device.
+    def disconnect(self):
+        if not self._conn_handle:
+            return
+        self._ble.gap_disconnect(self._conn_handle)
+        self._reset()
+
+    # Issues an (asynchronous) read, will invoke callback with data.
+    def read(self, callback):
+        if not self.is_connected():
+            return
+        self._read_callback = callback
+        self._ble.gattc_read(self._conn_handle, self._value_handle)
+
+    # Sets a callback to be invoked when the device notifies us.
+    def irq(self, handler):
+        self._notify_callback = handler
+
+    def _update_value(self, data):
+        self._value = data
+        return self._value
+
+    def value(self):
+        return self._value
+
+    def write(self, data):
+        if self._value_handle_rx is not None and self._conn_handle is not None:
+            print('gattc_write({},{},{})'.format(self._conn_handle, self._value_handle, data))
```

Comment:
Do we still need this print? Or was it just for debugging?

Reply:
Remnant from debugging, thanks